packages/pyroscope-flamegraph/src/format/format.ts

```diff
@@ -90,6 +91,73 @@ class DurationFormatter {
   }
 }
 
+// this is a class and not a function because we can save some time by
+//   precalculating divider and suffix and not doing it on each iteration
+class SubSecondDurationFormatter {
+  divider = 1;
+
+  suffix = 'second';
+
+  durations: [number, string][] = [
+    [60, 'minute'],
+    [60, 'hour'],
+    [24, 'day'],
+    [30, 'month'],
+    [12, 'year'],
+  ];
+
+  subSecondDurations: [number, string][] = [
+    [1000, 'ms'],
+    [1000, 'μs'],
+  ];
+
+  units = '';
+
+  constructor(maxDur: number, units?: string) {
+    this.units = units || '';
+    // eslint-disable-next-line no-plusplus
+    for (let i = 0; i < this.durations.length; i++) {
+      const level = this.durations[i];
+      if (!level) {
+        console.warn('Could not calculate level');
+        break;
+      }
+
+      if (maxDur >= level[0]) {
+        this.divider *= level[0];
+        maxDur /= level[0];
+        // eslint-disable-next-line prefer-destructuring
+        this.suffix = level[1];
+      } else {
+        break;
+      }
+    }
+  }
+
+  format(samples: number, sampleRate: number): string {
+    let n = samples / sampleRate / this.divider;
+
+    if (n && !Number.isInteger(n) && this.divider === 1) {
```

Comment:
This logic is hard to understand, and it is faulty (see test cases above). I would not invest any more time into it.

I think you could use existing `DurationFormatter`, but modify it in two ways:
* Add milliseconds and microseconds to `durations` array, with `1000` multipliers each.
* Modify line 138 to be `let n = samples / sampleRate * 1000000 / this.divider` to compensate for extra precision.

And the rest of it should work just fine. 

@StasDachinsky let me know what you think.

Reply:
And by the way, if that's the case (and these are the only changes we're making to this formatter) we probably don't need a separate class and instead we could have a constructor argument in the existing `DurationFormatter`.

/cc @Rperry2174 — let us know if there's more changes that need to happen. E.g jaeger does [double numbers](https://github.com/jaegertracing/jaeger-ui/blob/e2ae62af03e820f261713cb32ba0290c734e34f8/packages/jaeger-ui/src/utils/date.tsx#L109-L112)

Reply:
Ok, I will update the code according to your recommendations

Reply:
@petethepig Can you please explain what you mean by the constructor argument for "DurationFormatter"? Should I keep the current logic when the suffix is selected by the constructor argument "maxDur", or do I need to update it to allow "DurationFormatter" to select the suffix based on the value passed in the format function (like in the "jaeger"), but only for "trase_samples"?

Reply:
@StasDachinsky by that I meant that we'd add a new argument to `DurationFormatter` constructor, something like `enableSubsecondPrecision`, and then if you want to include sub-second precision you'd call it with that extra boolean, e.g:

`new DurationFormatter(123, '', true)`

This way we can make it so that in all other cases `DurationFormatter` is still working the same way it used to (with no sub-second precision) and it only does work with sub-second precision in tracing mode.